src/Http/Http.Extensions/src/EndpointParameterMetadataContext.cs

```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Http.Metadata;
+
+/// <summary>
+/// Represents the information accessible during endpoint creation by types that implement <see cref="IEndpointParameterMetadataProvider"/>.
+/// </summary>
+public class EndpointParameterMetadataContext
+{
+    /// <summary>
+    /// Gets the parameter of the route handler delegate of the endpoint being created.
+    /// </summary>
+    public ParameterInfo? Parameter { get; init; }
+
+    /// <summary>
+    /// Gets the <see cref="MethodInfo"/> associated with the current route handler.
+    /// </summary>
+    public IServiceProvider? Services { get; init; }
+
+    /// <summary>
+    /// Gets the list of objects that will be added to the metadata of the endpoint.
+    /// </summary>
+    public IList<object> EndpointMetadata { get; init; } = new List<object>();
```

Comment:
The default value shouldn't be needed right?

Reply:
you can't have it be non-nullable and not set an initial value. That's why it was nullable before 😄 
<img width="885" alt="image" src="https://user-images.githubusercontent.com/249088/162279144-c8e10fb8-c506-4114-85a3-ed6308aa5a72.png">

Reply:
I guess that's one reason to keep using a constructor until we get https://github.com/dotnet/csharplang/issues/3630. I'm okay with doing something like the following:

```suggestion
    public IList<object> EndpointMetadata { get; init; } = null!;
```

It doesn't matter so much here, but `Parameter` should also be non-nullable. `Services` on the other hand can actually be null.